cpp/src/arrow/python/parquet_encryption.cc

```diff
@@ -0,0 +1,75 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/python/parquet_encryption.h"
+
+namespace arrow {
+namespace py {
+namespace parquet {
+namespace encryption {
+
+PyKmsClient::PyKmsClient(PyObject* handler, PyKmsClientVtable vtable)
+    : handler_(handler), vtable_(std::move(vtable)) {
+  Py_INCREF(handler);
+}
+
+PyKmsClient::~PyKmsClient() {}
+
+std::string PyKmsClient::WrapKey(const std::string& key_bytes,
+                                 const std::string& master_key_identifier) {
+  std::string wrapped;
+  auto st = SafeCallIntoPython([&]() -> Status {
+    vtable_.wrap_key(handler_.obj(), key_bytes, master_key_identifier, &wrapped);
+    return Status::OK();
```

Comment:
There is a problem here: how are errors handled? A Python method can raise an exception. It should be transformed somehow into a `Status` (ideally) or a Parquet exception (but it would then lose information about the original Python exception).

Reply:
If I understand correctly the SafeCallIntoPython(), it preserves the current thread's error status, so if the Python method raises an exception, it will be preserved and the Python caller will eventually get it.

Reply:
This is right, but the challenge here is for the C++ calling code to get notified of the error (otherwise it will happily move along with an empty string).

Reply:
(slightly related: https://github.com/apache/arrow/pull/10566)

Reply:
Hi @pitrou , thank you for your observation. 
We cannot return a `Status` because PyKmsClient has to implement the KmsClient interface.
So we changed the methods to throw a ParquetStatusException, which allows to preserve the Status with Python exception details.